packages/flutter/lib/src/material/menu_bar.dart

```diff
@@ -0,0 +1,3256 @@
+import 'package:flutter/services.dart';
+import 'package:flutter/widgets.dart';
+
+import 'button_style.dart';
+import 'color_scheme.dart';
+import 'divider.dart';
+import 'icons.dart';
+import 'material.dart';
+import 'material_localizations.dart';
+import 'material_state.dart';
+import 'menu_bar_theme.dart';
+import 'text_button.dart';
+import 'text_button_theme.dart';
+import 'theme.dart';
+import 'theme_data.dart';
+
+// The time after a menu item is opened via hover during which clicks on a child
+// menu item with a submenu will be ignored by the menu item. This is so that
+// users aren't surprised by hovering over a main menu to open it, and then
+// quickly moving to a submenu and clicking on it to open it. Without this
+// delay, it often just feels like the menu opened briefly and closed when
+// clicked on, when in reality they opened it via the hover before they ever
+// clicked on it.
+const Duration _kMenuHoverClickBanDelay = Duration(milliseconds: 500);
+
+// The default size of the arrow that indicates that a menu has a submenu.
+const double _kDefaultSubmenuIconSize = 24.0;
+
+// The default spacing between the the leading icon, label, trailing icon, and
+// shortcut label in a _MenuBarItemLabel.
+const double _kLabelItemDefaultSpacing = 18.0;
+
+// The minimum spacing between the the leading icon, label, trailing icon, and
+// shortcut label in a _MenuBarItemLabel.
+const double _kLabelItemMinSpacing = 4.0;
+
+const Map<ShortcutActivator, Intent> _kMenuTraversalShortcuts = <ShortcutActivator, Intent>{
+  SingleActivator(LogicalKeyboardKey.enter): ActivateIntent(),
+  SingleActivator(LogicalKeyboardKey.numpadEnter): ActivateIntent(),
+  SingleActivator(LogicalKeyboardKey.space): ActivateIntent(),
+  SingleActivator(LogicalKeyboardKey.gameButtonA): ActivateIntent(),
+  SingleActivator(LogicalKeyboardKey.escape): DismissIntent(),
+  SingleActivator(LogicalKeyboardKey.tab): NextFocusIntent(),
+  SingleActivator(LogicalKeyboardKey.tab, shift: true): PreviousFocusIntent(),
+  SingleActivator(LogicalKeyboardKey.arrowDown): DirectionalFocusIntent(TraversalDirection.down),
+  SingleActivator(LogicalKeyboardKey.arrowUp): DirectionalFocusIntent(TraversalDirection.up),
+  SingleActivator(LogicalKeyboardKey.arrowLeft): DirectionalFocusIntent(TraversalDirection.left),
+  SingleActivator(LogicalKeyboardKey.arrowRight): DirectionalFocusIntent(TraversalDirection.right),
+};
+
+/// A wrapper class for [MenuItem] that contains extra metadata that allows
+/// rendering of the menu item, including the parent and children for this node,
+/// forming a tree.
+class _MenuNode with Diagnosticable, DiagnosticableTreeMixin, Comparable<_MenuNode> {
+  _MenuNode({required this.item, this.parent}) : children = <_MenuNode>[] {
+    if (item.members.isNotEmpty) {
+      // Don't add groups to the parent, just the members of the group.
+      for (final MenuItem member in item.members) {
+        _MenuNode(item: member, parent: parent);
```

Comment:
I reorganized this so that you have to call `addChildren` after creating the node. It doesn't need any arguments, though, since all the information is in the `MenuItem`.